pkg/controllers/state/cluster.go

```diff
@@ -523,6 +525,44 @@ func (c *Cluster) populateInflight(ctx context.Context, n *StateNode) error {
 }
 
 func (c *Cluster) populateVolumeLimits(ctx context.Context, n *StateNode) error {
+	if err := c.addStaticVolumeLimits(ctx, n); err != nil {
+		return fmt.Errorf("adding static volume limits, %w", err)
+	}
+	if err := c.addDynamicVolumeLimits(ctx, n); err != nil {
+		return fmt.Errorf("adding dynamic volume limits, %w", err)
+	}
+	return nil
+}
+
+func (c *Cluster) addStaticVolumeLimits(ctx context.Context, n *StateNode) error {
```

Comment:
Do we get ourselves into an infinite scale-up here if the local static provisioner stops doing its job and we have pods that are requesting from a storage class that never succeeds to provisions PVs? From looking at the code here, I think we will probably be fine, because we just won't assume a limit without the presence of any PVs. Should we assume the limit value is 0 here? I'm assuming this really depends on how quickly the local static provisioner brings PVs up with respect to when the node joins.

Reply:
That’s a great point.. Looking back through, it does seem like this is held together by explicitly not setting a limit if there’s no PVs. So if a new node is spun up and the pod is waiting for PVs to get created, then Karpenter would continue to nominate that node because it has no limit (yet). 

In that case I think we’re safe from an infinite scale-up, but it does mean that a node where PVs fail to create would be continuously nominated

Reply:
True -- either the workload or NodePool would need to specify that requirement. I don't think it's _too_ dissimilar from things like external devices (e.g. if your GPU device plugin fails).

In my opinion, the missing piece is that `kube-scheduler` knows when a local-PV workload won't fit on a given set of nodes; whereas Karpenter doesn't know since they're currently ignored. 

Neither `kube-scheduler` nor Karpenter can be sure of whether a given node will eventually host local PVs, but they can semi-reasonably rule out the ones that are "full". In that case, a node that Karpenter continuously nominates for a local-PV workload could be considered "not-ready". 

It'd be best-effort, though how that's actually implemented in this PR definitely isn't very clear 😄

Reply:
True -- either the workload or NodePool would need to specify that requirement. I don't think it's too dissimilar from things like external devices (e.g. if your GPU device plugin fails).

In my opinion, the missing piece is that kube-scheduler knows when a local-PV workload won't fit on a given set of nodes; whereas Karpenter doesn't know since they're currently ignored.

Neither kube-scheduler nor Karpenter can be sure of whether a given node will eventually host local PVs, but they can semi-reasonably rule out the ones that are "full". In that case, a node that Karpenter continuously nominates for a local-PV workload could be considered "not-ready".

It'd be best-effort, though how that's actually implemented in this PR definitely isn't very clear 😄

Reply:
For total transparency, I think https://github.com/aws/karpenter-provider-aws/pull/4735 is the direction our use-case will ultimately want to head in anyway  but I have seen issue this brought up in other forums; understood if the hacky-ish implementation isn't justified here